cmd/agent_autopilot.go

```diff
@@ -151,6 +168,122 @@ var agentAutopilotCmd = &cobra.Command{
+}
+
+// needToPullFunds returns whether the payAmt is larger than the current
+// balance of the Operator wallet.
+func needToPullFunds(cmd *cobra.Command, payAmt *big.Int) (bool, error) {
+	lapi, closer, err := PoolsSDK.Extern().ConnectLotusClient()
+	if err != nil {
+		return false, fmt.Errorf("Failed to instantiate eth client %s", err)
+	}
+	defer closer()
+
+	agentAddr, _, _, err := commonOwnerOrOperatorSetup(cmd)
+	if err != nil {
+		return false, err
+	}
+
+	agentAddrEthType, err := ethtypes.ParseEthAddress(agentAddr.String())
+	if err != nil {
+		return false, err
+	}
+
+	agentAddrDel, err := agentAddrEthType.ToFilecoinAddress()
+	if err != nil {
+		return false, err
+	}
+
+	agentFILIDAddr, err := lapi.StateLookupID(cmd.Context(), agentAddrDel, types.EmptyTSK)
+	if err != nil {
+		return false, err
+	}
+
+	actor, err := lapi.StateGetActor(cmd.Context(), agentFILIDAddr, types.EmptyTSK)
+	if err != nil {
+		return false, err
+	}
+	bal := actor.Balance
+
+	return payAmt.Cmp(bal.Int) > 0, nil
+}
+
+func pullFundsFromMiner(cmd *cobra.Command, miner address.Address, amount *big.Int) error {
+	agentAddr, senderKey, requesterKey, err := commonOwnerOrOperatorSetup(cmd)
+	if err != nil {
+		return err
+	}
+	pullevt := journal.RegisterEventType("agent", "pull")
+	evt := &events.AgentMinerPull{
+		AgentID: agentAddr.String(),
+		MinerID: miner.String(),
+		Amount:  amount.String(),
+	}
+	defer journal.RecordEvent(pullevt, func() interface{} { return evt })
+
+	tx, err := PoolsSDK.Act().AgentPullFunds(cmd.Context(), agentAddr, amount, miner, senderKey, requesterKey)
+	if err != nil {
+		evt.Error = err.Error()
+		return err
+	}
+	evt.Tx = tx.Hash().String()
+
+	// transaction landed on chain or errored
+	_, err = PoolsSDK.Query().StateWaitReceipt(cmd.Context(), tx.Hash())
+	if err != nil {
+		evt.Error = err.Error()
+		return err
+	}
+	return nil
+}
+
+// chooseMiner returns the 'richest' miner associated with the agent and whether that miner
+// has more funds than the requiredFunds parameter.
+func chooseMiner(cmd *cobra.Command, requiredFunds *big.Int) (address.Address, bool, error) {
+	lapi, closer, err := PoolsSDK.Extern().ConnectLotusClient()
+	if err != nil {
+		return address.Address{}, false, fmt.Errorf("Failed to instantiate eth client %s", err)
+	}
+	defer closer()
+	agentAddr, _, _, err := commonOwnerOrOperatorSetup(cmd)
+	if err != nil {
+		return address.Address{}, false, err
+	}
+
+	miners, err := PoolsSDK.Query().AgentMiners(cmd.Context(), agentAddr, nil)
+	if err != nil {
+		return address.Address{}, false, err
+	}
+
+	var chosen address.Address
+	var bal *big.Int
+	for i, m := range miners {
+		mbal, err := lapi.StateMinerAvailableBalance(cmd.Context(), m, types.EmptyTSK)
+		if err != nil {
+			return address.Address{}, false, err
+		}
+		if i == 0 {
+			chosen = m
+			bal = mbal.Int
+			continue
+		}
+		if mbal.Int.Cmp(bal) > 0 {
+			chosen = m
+			bal = mbal.Int
+		}
+	}
+	return chosen, bal.Cmp(requiredFunds) > 0, nil
+}
+
```

Comment:
I think we had both thought this was a good idea, but i'm realizing now that it may not be what SPs want - they may want to use a smaller miner that was pledged later to pay bills or something (for example).

Instead, lets just set a single miner-id in the autopilot config which is the single miner to pull funds from to make the payment. If that single miner doesn't have the funds, we throw the error. More explicit and less surprising things can happen that we have to try and explain later - let them choose.

If folks start asking for multi-pull miner config option, can add later

Reply:
I will leave the function there, but change out the current method to use the config value.